### PR TITLE
Add support for filtering service fields

### PR DESCRIPTION
**Pull request by bramkragten**
## Proposed change

Ads support for filtering service data fields based on filters:
https://github.com/home-assistant/core/pull/86162

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/data/selector.ts [hunk 1]
```diff
@@ -323,6 +324,50 @@ export interface UiColorSelector {
...
+  devices: HomeAssistant["devices"],
+  entities: HomeAssistant["entities"],
+  _targetSelector: TargetSelector
+) => {
+  const newEntities: string[] = [];
+  const newDevices: string[] = [];
+  Object.values(devices).forEach((device) => {
+    if (
+      device.area_id === areaId
+      // && deviceMeetsTargetFilter(device, targetSelector)
```
Comment (bramkragten) [inline src/data/selector.ts:338]: TODO

File: src/data/selector.ts [hunk 2]
```diff
@@ -323,6 +324,50 @@ export interface UiColorSelector {
...
+    ) {
+      newDevices.push(device.id);
+    }
+  });
+  Object.values(entities).forEach((entity) => {
+    if (
+      entity.area_id === areaId
+      // && entityMeetsTargetFilter(entity, targetSelector)
```
Comment (bramkragten) [inline src/data/selector.ts:346]: TODO

File: src/data/selector.ts [hunk 3]
```diff
@@ -323,6 +324,50 @@ export interface UiColorSelector {
...
+export const expandDeviceTarget = (
+  deviceId: string,
+  entities: HomeAssistant["entities"],
+  _targetSelector: TargetSelector
+) => {
+  const newEntities: string[] = [];
+  Object.values(entities).forEach((entity) => {
+    if (
+      entity.device_id === deviceId
+      // && entityMeetsTargetFilter(entity, targetSelector)
```
Comment (bramkragten) [inline src/data/selector.ts:363]: TODO

**Review by balloob — DISMISSED** [review]

File: src/data/selector.ts [hunk 4]
```diff
@@ -361,10 +366,121 @@ export interface UiColorSelector {
...
+        entitySources
+      )
+    ) {
+      newEntities.push(entity.entity_id);
+    }
+  });
+  return { devices: newDevices, entities: newEntities };
+};
+
+export const expandDeviceTarget = (
```
Comment (piitaya) [inline src/data/selector.ts:408]: May be we should move all these helpers to another file to separate the types and the helpers

**Review by piitaya — APPROVED** [review]

**Pull request merged**